src/bokeh/models/sources.py

```diff
@@ -507,7 +507,9 @@ def _stream(self, new_data: DataDict | pd.Series | pd.DataFrame,
                 in this case the Series index is used as column names
 
                 All columns of the data source must be present in ``new_data``,
-                with identical-length append data.
+                with identical-length append data. In case the data source has
+                no columns at all, the extra columns in ``new_data`` will not
+                raise an exception because of mismatched columns.
```

Comment:
This reads a bit too roundabout to me. Maybe something like this?
```suggestion
               All columns in ``new_data`` must be the same length. In case
               data source is not empty, the column names in ``new_data`` 
               must match those of the data source exactly (otherwise an
               exception is raised). If the data source is empty then 
               ``new_data`` replaces the empty ``.data`` dict of the
               data source.
  
```

Reply:
@bryevdv Much better :)